Apply is_ceiling/is_wall swap fix to 2D move_and_slide

20kdc opened the pull request:
This fixes is_ceiling being swapped with is_wall.

Essentially a condition was slightly the wrong way around, though I've gone and added some comments to that area to help understand the code, partially to help myself understand it,and partially in case of future maintainers.

This project (note: Do not press "3D" in the menu, as that is meant for testing another PR.) shows the bug on master.
[kTest.zip](https://github.com/godotengine/godot/files/922179/kTest.zip)

tagcup commented:
This explanation doesn't make no sense to me.

There is one condition: `surface_n.dot(collision_n) >= cos(threshold)`. You shouldn't be changing the direction of the inequality just because you're dealing with a different surface.

Meaning

`else if (get_collision_normal().dot(-p_floor_direction) <= Math::cos(p_floor_max_angle)) { //ceiling`

should simply be replaced by

`} else if (get_collision_normal().dot(-p_floor_direction) => Math::cos(p_floor_max_angle)) { //ceiling`

And I don't think this requires any further explanation.

tagcup commented:
I was in a hurry, so I forgot to mention a few points, sorry.

So basically, if you can express things in the vector form, you can write general expressions without worrying about the coordinate systems etc. (in general terms, if you have gradients etc, you need to make sure you're using covariant derivatives of course)

And `surface_n.dot(collision_n) >= cos(threshold)` is like that. It's a vector expression, and as long as you feed it the correct surface normal and collision normal, it will work.

In your PR, you're giving the wrong surface normal, and furthermore flipping the direction of the inequality condition. This double-sign mistake happens to give correct results, but that's of course not the proper way to fix it.

So could you please modify your PR such that it's just replacing that line with:

`else if (get_collision_normal().dot(-p_floor_direction) => Math::cos(p_floor_max_angle)) { //ceiling`

and remove the explanation you wrote?

20kdc commented:
Huh, really? I thought it made sense, but I'll change it.
I guess I don't think in... whatever system you're using.

20kdc committed:
Apply is_ceiling/is_wall swap fix to 2D move_and_slide (minus explaination)

As it turns out, is_ceiling would be true when hitting a wall,
 and is_wall would be true momentarily when hitting a ceiling.

This makes a tiny one-line change to fix that.
Without trying to explain the code for anyone else having to
 mess around with it.

tagcup commented:
The nice thing is when/if we add the ceiling normal as a distinct parameter, we can simply change that line to

`} else if (get_collision_normal().dot(p_ceiling_direction) >= Math::cos(p_floor_max_angle)) { //ceiling`

tagcup commented:
LGTM

20kdc commented:
@tagcup So, you're saying I should add the parameter in this PR? I'm a bit confused since that seems to go against PR policy, but alright

bojidar-bg commented:
@20kdc it is ["Looks Good To Me"](http://acronyms.thefreedictionary.com/LGTM)

tagcup commented:
Yep, I just meant "this PR in it's current state looks good to me"
